src/backend/largebuddyrange.h

```diff
@@ -0,0 +1,287 @@
+#pragma once
+
+#include "../ds/address.h"
+#include "../ds/bits.h"
+#include "../mem/allocconfig.h"
+#include "../mem/metaslab.h"
+#include "../pal/pal.h"
+#include "buddy.h"
+#include "range_helpers.h"
+
+#include <string>
+
+namespace snmalloc
+{
+  /**
+   * Class for using the pagemap entries for the buddy allocator.
+   */
+  template<typename Pagemap>
+  class BuddyChunkRep
+  {
+  public:
+    using Holder = uintptr_t;
+    using Contents = uintptr_t;
+
+    static constexpr uintptr_t RED_BIT = 2;
+
+    static constexpr Contents null = 0;
+
+    static void set(Holder* ptr, Contents r)
+    {
+      SNMALLOC_ASSERT((r & (MIN_CHUNK_SIZE - 1)) == 0);
+      // Preserve lower bits.
+      *ptr = r | (*ptr & (MIN_CHUNK_SIZE - 1)) | BACKEND_MARKER;
```

Comment:
The CHERI compiler is again a little upset about this expression: `*ptr` is a capability, `*ptr & ...` is a capability, and `r` is a capability, so `r | (*ptr & ...)` again triggers its uncertainty of provenance.
```suggestion
      *ptr = r | statc_cast<address_t>(*ptr & (MIN_CHUNK_SIZE - 1)) | BACKEND_MARKER;
```